File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/plugin/js/JSProduct.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

package com.mytiki.sdk.capture.receipt.capacitor.plugin.js

import com.getcapacitor.JSObject
import com.microblink.core.Product
import org.json.JSONArray
import org.json.JSONObject

/**
 * Represents a parsed product from a receipt using the RSP (Receipt Structure Parser) framework.
 *
 * @property productNumber The product number, if available.
 * @property description The product description, if available.
 * @property quantity The quantity of the product, if available.
 * @property unitPrice The unit price of the product, if available.
 * @property unitOfMeasure The unit of measure for the product, if available.
 * @property totalPrice The total price of the product, if available.
 * @property fullPrice The full price of the product.
 * @property line The line number of the product on the receipt.
 * @property productName The product name, if available.
 * @property brand The product brand, if available.
 * @property category The product category, if available.
 * @property size The product size, if available.
 * @property rewardsGroup The rewards group associated with the product, if available.
 * @property competitorRewardsGroup The competitor's rewards group, if available.
 * @property upc The Universal Product Code (UPC) of the product, if available.
 * @property imageUrl The URL of the product image, if available.
 * @property shippingStatus The shipping status of the product, if available.
 * @property additionalLines Additional lines of text associated with the product.
 * @property priceAfterCoupons The product price after applying coupons, if available.
 * @property voided Indicates whether the product is voided.
 * @property probability The probability score associated with the product.
 * @property sensitive Indicates whether the product is sensitive.
 * @property possibleProducts List of possible sub-products associated with the product.
 * @property subProducts List of sub-products associated with the product.
 * @property added Indicates whether the product was added.
 * @property blinkReceiptBrand The brand information from the BlinkReceipt SDK, if available.
 * @property blinkReceiptCategory The category information from the BlinkReceipt SDK, if available.
 * @property extendedFields Additional extended fields associated with the product.
 * @property fuelType The fuel type of the product, if available.
 * @property descriptionPrefix The prefix of the product description, if available.
 * @property descriptionPostfix The postfix of the product description, if available.
 * @property skuPrefix The prefix of the product SKU, if available.
 * @property skuPostfix The postfix of the product SKU, if available.
 * @property attributes List of attributes associated with the product.
 * @property sector The sector of the product, if available.
 * @property department The department of the product, if available.
 * @property majorCategory The major category of the product, if available.
 * @property subCategory The sub-category of the product, if available.
 * @property itemType The item type of the product, if available.
 */
class JSProduct(product: Product) {
    private val productNumber: JSStringType?
    private val description: JSStringType?
    private val quantity: JSFloatType?
    private val unitPrice: JSFloatType?
    private val unitOfMeasure: JSStringType?
    private val totalPrice: JSFloatType?
    private val fullPrice: Float
    private val line: Int
    private val productName: String?
    private val brand: String?
    private val category: String?
    private val size: String?
    private val rewardsGroup: String?
    private val competitorRewardsGroup: String?
    private val upc: String?
    private val imageUrl: String?
    private val shippingStatus: String?
    private val additionalLines: List<JSAdditionalLine>
    private val priceAfterCoupons: JSFloatType?
    private val voided: Boolean
    private val probability: Double
    private val sensitive: Boolean
    private val possibleProducts: List<JSProduct>
    private val subProducts: List<JSProduct>
    private val added: Boolean
    private val blinkReceiptBrand: String?
    private val blinkReceiptCategory: String?
    private val extendedFields: JSONObject?
    private val fuelType: String?
    private val descriptionPrefix: JSStringType?
    private val descriptionPostfix: JSStringType?
    private val skuPrefix: JSStringType?
    private val skuPostfix: JSStringType?
    private val attributes: List<JSONObject>
    private val sector: String?
    private val department: String?
    private val majorCategory: String?
    private val subCategory: String?
    private val itemType: String?

    init {
        productNumber = JSStringType.opt(product.productNumber())
        description = JSStringType.opt(product.description())
        quantity = JSFloatType.opt(product.quantity())
        unitPrice = JSFloatType.opt(product.unitPrice())
        unitOfMeasure = JSStringType.opt(product.unitOfMeasure())
        totalPrice = JSFloatType.opt(product.totalPrice())
        fullPrice = product.fullPrice()
        line = product.line()
        productName = product.productName()
        brand = product.brand()
        category = product.category()
        size = product.size()
        rewardsGroup = product.rewardsGroup()
        competitorRewardsGroup = product.competitorRewardsGroup()
        upc = product.upc()
        imageUrl = product.imageUrl()
        shippingStatus = product.shippingStatus()
        additionalLines =
            product.additionalLines()?.map { additionalLine -> JSAdditionalLine(additionalLine) }
                ?: emptyList()
        priceAfterCoupons = JSFloatType.opt(product.priceAfterCoupons())
        voided = product.voided()
        probability = product.probability()
        sensitive = product.sensitive()
        possibleProducts = product.possibleProducts()?.map { prd -> JSProduct(prd) } ?: emptyList()
        subProducts = product.subProducts()?.map { prd -> JSProduct(prd) } ?: emptyList()
        added = product.added()
        blinkReceiptBrand = product.blinkReceiptBrand()
        blinkReceiptCategory = product.blinkReceiptCategory()
        fuelType = product.fuelType()
        descriptionPrefix = JSStringType.opt(product.descriptionPrefix())
        descriptionPostfix = JSStringType.opt(product.skuPostfix())
        skuPrefix = JSStringType.opt(product.skuPrefix())
        skuPostfix = JSStringType.opt(product.skuPostfix())
        sector = product.sector()
        department = product.department()
        majorCategory = product.majorCategory()
        subCategory = product.subCategory()
        itemType = product.itemType()
        extendedFields = if (product.extendedFields() != null) {
            val extendedFields = JSObject()
            product.extendedFields()
                ?.forEach { entry -> extendedFields.put(entry.key, entry.value) }
            extendedFields
        } else null
        attributes = if (product.attributes() != null) {
            product.attributes()!!.map { attr ->
                val json = JSObject()
                attr.forEach { entry -> json.put(entry.key, entry.value) }
                json
            }
        } else emptyList()
    }

    /**
     * Converts the RSP product to a JSON representation.
     *
     * @return A JSON object representing the RSP product.
     */
    fun toJS(): JSObject =
        JSObject()
            .put("productNumber", productNumber?.toJS())
            .put("description", description?.toJS())
            .put("quantity", quantity?.toJS())
            .put("unitPrice", unitPrice?.toJS())
            .put("unitOfMeasure", unitOfMeasure?.toJS())
            .put("totalPrice", totalPrice?.toJS())
            .put("fullPrice", fullPrice)
            .put("line", line)
            .put("productName", productName)
            .put("brand", brand)
            .put("category", category)
            .put("size", size)
            .put("rewardsGroup", rewardsGroup)
            .put("competitorRewardsGroup", competitorRewardsGroup)
            .put("upc", upc)
            .put("imageUrl", imageUrl)
            .put("shippingStatus", shippingStatus)
            .put("additionalLines", JSONArray(additionalLines.map { line -> line.toJS() }))
            .put("priceAfterCoupons", priceAfterCoupons?.toJS())
            .put("voided", voided)
            .put("probability", probability)
            .put("sensitive", sensitive)
            .put("possibleProducts", JSONArray(possibleProducts.map { prd -> prd.toJS() }))
            .put("subProducts", JSONArray(subProducts.map { prd -> prd.toJS() }))
            .put("added", added)
            .put("blinkReceiptBrand", blinkReceiptBrand)
            .put("blinkReceiptCategory", blinkReceiptCategory)
            .put("extendedFields", extendedFields)
            .put("fuelType", fuelType)
            .put("descriptionPrefix", descriptionPrefix?.toJS())
            .put("descriptionPostfix", descriptionPostfix?.toJS())
            .put("skuPrefix", skuPrefix?.toJS())
            .put("skuPostfix", skuPostfix?.toJS())
            .put("attributes", JSONArray(attributes))
            .put("sector", sector)
            .put("department", department)
            .put("majorCategory", majorCategory)
            .put("subCategory", subCategory)
            .put("itemType", itemType)

    companion object {
        /**
         * Creates an optional [JSProduct] from a [Product] instance.
         *
         * @param product The [Product] instance to create an [JSProduct] from.
         * @return An [JSProduct] instance or null if the input is null.
         */
        fun opt(product: Product?): JSProduct? =
            if (product != null) JSProduct(product) else null
    }
}

File: ios/Plugin/Plugin/Js/JSProduct.swift
/*
 * RspProduct Struct
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import BlinkReceipt
import BlinkEReceipt
import Capacitor

/**
 Represents a response containing product information.

 This struct is used to convey product details, including product number, description, quantity, pricing, brand, category, and more.
 */
struct JSProduct {
    /// The product number, if available.
    private let productNumber: JSStringType?
    /// The product description, if available.
    private let description: JSStringType?
    /// The quantity of the product, if available.
    private let quantity: JSFloatType?
    /// The unit price of the product, if available.
    private let unitPrice: JSFloatType?
    /// The unit of measure for the product, if available.
    private let unitOfMeasure: JSStringType?
    /// The total price of the product, if available.
    private let totalPrice: JSFloatType?
    /// The full price of the product, if available.
    private let fullPrice: JSFloatType
    /// The product name, if available.
    private let productName: String?
    /// The brand of the product, if available.
    private let brand: String?
    /// The category of the product, if available.
    private let category: String?
    /// The size of the product, if available.
    private let size: String?
    /// The UPC code for the product
    private let upc: String?
    /// The URL of the product image, if available.
    private let imageUrl: String?
    /// The shipping status of the product, if available.
    private let shippingStatus: String?
    /// Additional lines associated with the product.
    private let additionalLines: [JSAdditionalLine]?
    /// The price of the product after applying coupons, if available.
    private let priceAfterCoupons: JSFloatType?
    /// Indicates if the product is voided.
    private let voided: Bool?
    /// The probability associated with the product.
    private let probability: Float?
    /// Indicates if the product is sensitive.
    private let sensitive: Bool?
    /// Possible sub-products associated with the product.
    private let possibleProducts: [JSProduct]
    /// Sub-products of the product.
    private let subProducts: [JSProduct]
    /// Indicates if the product is user-added.
    private let added: Bool?
    /// The brand according to BlinkReceipt.
    private let blinkReceiptBrand: String?
    /// The category according to BlinkReceipt.
    private let blinkReceiptCategory: String?
    /// A map of extended fields associated with the product.
    private let extendedFields: JSObject?
    /// The fuel type of the product.
    private let fuelType: String?
    /// The description prefix, if available.
    private let descriptionPrefix: JSStringType?
    /// The description postfix, if available.
    private let descriptionPostfix: JSStringType?
    /// The SKU prefix, if available.
    private let skuPrefix: JSStringType?
    /// The SKU postfix, if available.
    private let skuPostfix: JSStringType?
    /// The sector of the product.
    private let sector: String?
    /// The department of the product.
    private let department: String?
    /// The major category of the product.
    private let majorCategory: String?
    /// The sub-category of the product.
    private let subCategory: String?
    /// The item type of the product.
    private let itemType: String?

    /**
     Initializes an `RspProduct` struct.

     - Parameters:
        - product: The `BRProduct` object containing product information.
     */
    init(product: BRProduct) {
        productNumber = JSStringType(stringType: product.productNumber)
        description = JSStringType(string: product.description)
        quantity = JSFloatType(floatType: product.quantity)
        unitPrice = JSFloatType(floatType: product.unitPrice)
        unitOfMeasure = JSStringType(stringType: product.unitOfMeasure)
        totalPrice = JSFloatType(floatType: product.totalPrice)
        fullPrice = JSFloatType(floatType: product.fullPrice)
        productName = product.productName
        brand = product.brand
        category = product.category
        size = product.size
        upc = product.upc
        imageUrl = product.imgUrl
        shippingStatus = product.shippingStatus
        additionalLines = product.additionalLines?.map { additionalLine in JSAdditionalLine(additionalLine: additionalLine) } ?? []
        priceAfterCoupons = JSFloatType(floatType: product.priceAfterCoupons)
        voided = product.isVoided
        probability = product.probability
        sensitive = product.isSensitive
        possibleProducts = product.possibleProducts?.map { prd in JSProduct(product: prd) } ?? []
        subProducts = product.subProducts?.map { prd in JSProduct(product: prd) } ?? []
        added = product.userAdded
        blinkReceiptBrand = product.brand
        blinkReceiptCategory = product.category
        extendedFields = product.extendedFields != nil ? {
            var ret = JSObject()
            product.extendedFields!.forEach { (key: AnyHashable, value: Any) in
                ret.updateValue(value as! any JSValue, forKey: key as! String)
            }
            return ret
        }() : nil
        fuelType = product.fuelType
        descriptionPrefix = JSStringType(stringType: product.prodDescPrefix)
        descriptionPostfix = JSStringType(stringType: product.prodDescPostfix)
        skuPrefix = JSStringType(stringType: product.prodNumPrefix)
        skuPostfix = JSStringType(stringType: product.prodNumPostfix)
        sector = product.sector
        department = product.department
        majorCategory = product.majorCategory
        subCategory = product.subCategory
        itemType = product.itemType
    }

    /**
     Converts the `RspProduct` struct into a dictionary suitable for use in plugin response data.

     - Returns: A dictionary containing product information in a format suitable for a Capacitor plugin call result.
     */
    func toJSObject() -> JSObject {
        var ret = JSObject()
        ret["productNumber"] = productNumber?.toJSObject()
        ret["description"] = description?.toJSObject()
        ret["quantity"] = quantity?.toJSObject()
        ret["unitPrice"] = unitPrice?.toJSObject()
        ret["unitOfMeasure"] = unitOfMeasure?.toJSObject()
        ret["totalPrice"] = totalPrice?.toJSObject()
        ret["fullPrice"] = fullPrice.toJSObject()
        ret["productName"] = productName
        ret["brand"] = brand
        ret["category"] = category
        ret["size"] = size
        ret["imageUrl"] = imageUrl
        ret["shippingStatus"] = shippingStatus
        ret["additionalLines"] = additionalLines
        ret["priceAfterCoupons"] = priceAfterCoupons?.toJSObject()
        ret["voided"] = voided
        ret["probability"] = probability
        ret["sensitive"] = sensitive
        ret["possibleProducts"] = possibleProducts
        ret["subProducts"] = subProducts
        ret["added"] = added
        ret["blinkReceiptBrand"] = blinkReceiptBrand
        ret["blinkReceiptCategory"] = blinkReceiptCategory
        ret["fuelType"] = fuelType
        ret["descriptionPrefix"] = descriptionPrefix?.toJSObject()
        ret["descriptionPostfix"] = descriptionPostfix?.toJSObject()
        ret["skuPrefix"] = skuPrefix?.toJSObject()
        ret["skuPostfix"] = skuPostfix?.toJSObject()
        ret["sector"] = sector
        ret["department"] = department
        ret["majorCategory"] = majorCategory
        ret["subCategory"] = subCategory
        ret["itemType"] = itemType
        return ret
    }
}
